### ISSUE TITLE
Segmentation fault in Core::Shutdown

### ISSUE THREAD

**Issue by dParadiz** [libRocket/libRocket#247]
When trying to abstract librocket calling Core::Shutdown outside initialization function results in segmentation fault. Using x86_64 librocket with SDL2.

Used code

**RocketApp.h**:
```
#ifndef ROCKETAPP_H_INCLUDED
#define  ROCKETAPP_H_INCLUDED


#include <SDL2/SDL.h>
#include <string>

#include <Rocket/Core.h>
#include <Rocket/Controls.h>
#include <Rocket/Debugger.h>

class RocketApp {
public:
    RocketApp() {};
    virtual ~RocketApp() {};
    void init(SDL_Window* screen, SDL_Renderer* renderer, uint32_t width, uint32_t height);
    void shutDown();
private:
    Rocket::Core::Context* context;
};

#endif // 
```

**RocketApp.cpp**:
```
#include "RocketApp.h"

#include "SystemInterfaceSDL2.h"
#include "RenderInterfaceSDL2.h"
#include "ShellFileInterface.h"

void RocketApp::init(SDL_Window* screen, SDL_Renderer* renderer, uint32_t width, uint32_t height)
{
    RocketSDL2Renderer Renderer(renderer, screen);
    RocketSDL2SystemInterface SystemInterface;
    ShellFileInterface FileInterface("assets/");

    Rocket::Core::SetFileInterface(&FileInterface);
    Rocket::Core::SetRenderInterface(&Renderer);
    Rocket::Core::SetSystemInterface(&SystemInterface);

    if(!Rocket::Core::Initialise()) {
       return;
    }

    Rocket::Core::FontDatabase::LoadFontFace("Delicious-Bold.otf");
    Rocket::Core::FontDatabase::LoadFontFace("Delicious-BoldItalic.otf");
    Rocket::Core::FontDatabase::LoadFontFace("Delicious-Italic.otf");
    Rocket::Core::FontDatabase::LoadFontFace("Delicious-Roman.otf");

    this->context = Rocket::Core::CreateContext("default", Rocket::Core::Vector2i(width, height));
    Rocket::Core::ElementDocument *Document = this->context->LoadDocument("demo.rml");

    Rocket::Debugger::Initialise(this->context);

    if(Document) {
        Document->Show();
        Document->RemoveReference();
    }
    //no segmentation fault
    //this->context->RemoveReference();
    //Rocket::Core::Shutdown();
}

void RocketApp::shutDown()
{
   if (this->context != NULL) {
        //segmentation fault if called here
        this->context->RemoveReference();
        Rocket::Core::Shutdown();
    }
}
```
**Stack trace**
> #0 0x7ffff73093bf	Rocket::Core::ElementUtilities::GetLineHeight(Rocket::Core::Element*) () 
> #1 0x7ffff732b79a	Rocket::Core::LayoutInlineBox::LayoutInlineBox(Rocket::Core::Element*, Rocket::Core::Box const&) () 
> #2 0x7ffff732c47c	Rocket::Core::LayoutInlineBoxText::LayoutInlineBoxText(Rocket::Core::Element*, int) () 
> #3 0x7ffff732d3af	Rocket::Core::LayoutLineBox::AddElement(Rocket::Core::Element*, Rocket::Core::Box const&) () 
> #4 0x7ffff7328270	Rocket::Core::LayoutBlockBox::AddInlineElement(Rocket::Core::Element*, Rocket::Core::Box const&) () 
> #5 0x7ffff732824f	Rocket::Core::LayoutBlockBox::AddInlineElement(Rocket::Core::Element*, Rocket::Core::Box const&) () 
> #6 0x7ffff732ad8f	Rocket::Core::LayoutEngine::FormatElementInline(Rocket::Core::Element*) () 
> #7 0x7ffff732af1b	Rocket::Core::LayoutEngine::FormatElement(Rocket::Core::Element*) () 
> #8 0x7ffff732afb5	Rocket::Core::LayoutEngine::FormatElement(Rocket::Core::Element*, Rocket::Core::Vector2<float> const&) () 
> #9 0x7ffff732aecc	Rocket::Core::LayoutEngine::FormatElement(Rocket::Core::Element*) () 
> #10 0x7ffff732afb5	Rocket::Core::LayoutEngine::FormatElement(Rocket::Core::Element*, Rocket::Core::Vector2<float> const&) () 
> #11 0x7ffff72f65af	Rocket::Core::ElementDocument::_UpdateLayout() () 
> #12 0x7ffff72b0df4	Rocket::Core::Context::GetElementAtPoint(Rocket::Core::Vector2<float> const&, Rocket::Core::Element const*, Rocket::Core::Element*) () 
> #13 0x7ffff72b680e	Rocket::Core::Context::UpdateHoverChain(Rocket::Core::Dictionary const&, Rocket::Core::Dictionary const&, Rocket::Core::Vector2<int> const&) () 
> #14 0x7ffff72b7223	Rocket::Core::Context::UnloadDocument(Rocket::Core::ElementDocument*) () 
> #15 0x7ffff70355db	Rocket::Debugger::Plugin::SetContext(Rocket::Core::Context*) () (vendor/builds/Rocket/lib/x86_64-linux-gnu/libRocketDebugger.so.1:??)
> #16 0x7ffff7038c4f	Rocket::Debugger::Plugin::OnContextDestroy(Rocket::Core::Context*) () (vendor/builds/Rocket/lib/x86_64-linux-gnu/libRocketDebugger.so.1:??)
> #17 0x7ffff732e4e9	Rocket::Core::PluginRegistry::NotifyContextDestroy(Rocket::Core::Context*) () 
> #18 0x7ffff72b73fa	Rocket::Core::Context::~Context() () 
> #19 0x7ffff72b7749	Rocket::Core::Context::~Context() () 
 >#20 0x406768	RocketApp::shutDown(this=0xcb1e30) (src/test.cpp:43)
> #21 0x40387d	main() (src/main.cpp:62)
> #19 0x7ffff72b7749	Rocket::Core::Context::~Context() () (vendor/builds/Rocket/lib/x86_64-linux-gnu/libRocketCore.so.1:??)
> #20 0x406768	RocketApp::shutDown(this=0xcb1e30) (src/test.cpp:43)
> #21 0x40387d	main() (src/main.cpp:62)

**Comment by dwimsey** [comment]
If you don’t release the context, does it issue any warnings or exit cleanly?

**Comment by dParadiz** [comment]
There are no warrnings just segmentation fault. And following stack trace

>#0 0x7ffff72bacce	Rocket::Core::SetSystemInterface(Rocket::Core::SystemInterface*) 
>#1 0x7ffff7039d60	Rocket::Debugger::SystemInterface::~SystemInterface()
>#2 0x7ffff7039d99	Rocket::Debugger::SystemInterface::~SystemInterface()
>#3 0x7ffff7038ab8	Rocket::Debugger::Plugin::ReleaseElements()
>#4 0x7ffff7038b0b	Rocket::Debugger::Plugin::OnShutdown() 
>#5 0x7ffff732e412	Rocket::Core::PluginRegistry::NotifyShutdown()
>#6 0x7ffff72baf12	Rocket::Core::Shutdown() 
>#7 0x406790	RocketApp::shutDown(this=0xcbbd80)
>#8 0x40387d	main()

**Comment by dwimsey** [comment]
Sounds like you’ve got some stack corruption/smashing going on somewhere.  My first guess would be to verify that all the libraries you are linking with are .so and not .a files.  9 times out of 10, a random problem like this in a tiny test app (assuming none of the other classes you provided (sys interface, render interface, file interface) are broken) is that you’ve linked one part of your app with the static version of a library (Freetype is a good example) and another part of your app with the dynamic version.  Things tend to go south quickly in random, obscure ways in these situations due to the memory allocator getting foobarred.  Thats easy to verify.

Have you compiled the included sample app and ran it?

**Comment by dParadiz** [comment]
I checked linked libraries and theye are pointing to .so libs. Thous that are directly included into build (Rocket*, libSDL2, lSDL2_image, glew, GL). In ccmake lib for Freetype is also set to .so file.

sys interface, render interface, file interface are thous provided with SDL2 sample. Basically all the code is taken from sample. Just Rocket part was moved out of the main function in separate class. 

While sample code was working moving Rocket code to separate class was causing this issue.
